### PR TITLE
feat: P4ADEV-2047 ExportFile download API

### PR DESCRIPTION
**Pull request by pelliccm**
#### How Has This Been Tested?

- Pre-Deploy Test
  - [x] Unit
  - [ ] Integration (Narrow)
- Post-Deploy Test
  - [ ] Isolated Microservice
  - [ ] Broader Integration
  - [ ] Acceptance
  - [ ] Performance & Load

#### Types of changes

- [ ] PATCH - Bug fix (backwards compatible bug fixes)
- [x] MINOR - New feature (add functionality in a backwards compatible manner)
- [ ] MAJOR - Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] CHORE - Minor Change (fix or feature that don't impact the functionality e.g. Documentation or lint configuration)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/client/ExportFileClient.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    ProcessExecutionsApisHolder processExecutionsApisHolder) {
+    this.processExecutionsApisHolder = processExecutionsApisHolder;
+  }
+
+  public ExportFile getExportFile(Long exportFileId, String accessToken) {
+    try {
+      log.debug("Fetching export file with ID [{}]", exportFileId);
+      return processExecutionsApisHolder.getExportFileEntityControllerApi(accessToken).crudGetExportfile(String.valueOf(exportFileId));
+    } catch (HttpClientErrorException.NotFound e) {
+      log.info("Cannot find ExportFile with ID [{}]", exportFileId, e);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/client/ExportFileClient.java]: don't print the stack trace! it has not value, we already know the error (maybe you could print just the message...)
```suggestion
      log.info("Cannot find ExportFile with ID [{}]", exportFileId);
```

File: src/main/java/it/gov/pagopa/pu/fileshare/controller/ExportFilesController.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  public ExportFilesController(
+    ExportFileFacadeService exportFileFacadeService) {
+    this.exportFileFacadeService = exportFileFacadeService;
+  }
+
+
+  @Override
+  public ResponseEntity<Resource> downloadExportFile(Long organizationId, Long exportFileId) {
+    FileResourceDTO fileResourceDTO = exportFileFacadeService.downloadExportFile(organizationId, exportFileId, SecurityUtils.getLoggedUser(), SecurityUtils.getAccessToken());
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/controller/ExportFilesController.java]: let's always log at the start of each API method

File: src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    this.archivedSubFolder = archivedSubFolder;
+  }
+
+  @Override
+  public FileResourceDTO downloadExportFile(Long organizationId,
+    Long exportFileId, UserInfo user,
+    String accessToken) {
+    userAuthorizationService.checkUserAuthorization(organizationId, user, accessToken);
+
+    ExportFile exportFile = exportFileClient.getExportFile(exportFileId, accessToken);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java]: don't directly call a Client class from business logic layer, expose the method through a Service on the connector layer

File: src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/client/ExportFileClient.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+@Service
+public class ExportFileClient {
+  private final ProcessExecutionsApisHolder processExecutionsApisHolder;
+
+  public ExportFileClient(
+    ProcessExecutionsApisHolder processExecutionsApisHolder) {
+    this.processExecutionsApisHolder = processExecutionsApisHolder;
+  }
+
+  public ExportFile getExportFile(Long exportFileId, String accessToken) {
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/client/ExportFileClient.java:19]: We should check if you we need to send also the current mappedExternalUserId in order to retrieve only current user file

File: src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  private Path getFilePath(ExportFile exportFile) {
+    Path organizationBasePath = fileStorerService.buildOrganizationBasePath(exportFile.getOrganizationId());
+
+    Path filePath = organizationBasePath
+      .resolve(exportFile.getFilePathName());
+    if (exportFile.getStatus() == COMPLETED || exportFile.getStatus() == ERROR) {
+      filePath = filePath
+        .resolve(archivedSubFolder);
+    }
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java]: exports doesn't have an archive folder! exports are stored only the configured path
(ingestionFlowFile are moved after their elaboration)

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  @Override
+  public ExportFile getExportFile(Long exportFileId, Long organizationId, UserInfo loggedUser, String accessToken) {
+    boolean isLoggedUserAdmin = AuthorizationService.isAdminRole(
+      organizationId, loggedUser);
+    String operatorExternalUserId = null;
+
+    if(!isLoggedUserAdmin){
+      operatorExternalUserId = loggedUser.getMappedExternalUserId();
+    }
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java]: This are business layer logic, not connector logic.
Move into ExportFileFacadeServiceImpl

File: src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    ExportFile exportFile = client.getExportFile(exportFileId, accessToken);
+
+    if (!isLoggedUserAdmin && !operatorExternalUserId.equals(exportFile.getOperatorExternalId())) {
+      throw new UnauthorizedFileDownloadException("User is not authorized to download export file with ID " + exportFileId);
+    }
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java]: also this is a business layer logic

File: src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java (hunk 6 above)
Comment (antonioT90) [reply src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java]: ps: this layer could be configured to introduce a cache! thus we cannot add business logic here

File: src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java (hunk 7 above)
Comment (antonioT90) [reply src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java]: ps: if you don't plan to pass the operatorExternalId to the client, why are you replicating the isLoggedUserAdmin check? The if at row 25 and this if are redundant...

Comment (antonioT90) [reply src/main/java/it/gov/pagopa/pu/fileshare/connector/processexecutions/ExportFileServiceImpl.java]: pps: exportFile could be null

File: src/main/java/it/gov/pagopa/pu/fileshare/controller/ExportFilesController.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+  public ExportFilesController(
+    ExportFileFacadeService exportFileFacadeService) {
+    this.exportFileFacadeService = exportFileFacadeService;
+  }
+
+
+  @Override
+  public ResponseEntity<Resource> downloadExportFile(Long organizationId, Long exportFileId) {
+    log.debug("Requesting to download export file [exportFileId: {}] of organization [organizationId: {}]", exportFileId, organizationId);
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/controller/ExportFilesController.java]:
```suggestion
    log.info("Requesting to download export file [exportFileId: {}] of organization [organizationId: {}]", exportFileId, organizationId);
```

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  @Override
+  public FileResourceDTO downloadExportFile(Long organizationId,
+    Long exportFileId, UserInfo user,
+    String accessToken) {
+    userAuthorizationService.checkUserAuthorization(organizationId, user,
+      accessToken);
+
+    ExportFile exportFile = exportFileService.getExportFile(exportFileId, accessToken);
+
+    if (!AuthorizationService.isAdminRole(organizationId, user) &&
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java:46]: If exportFile is null you should throw notFoundException (is this behavior not implemented for ingestionFlowFile? If not, please, let's fix it

If you add a test on with exportFile null you wil see that actually you will have a NullpointerException

**Review by antonioT90 — CHANGES_REQUESTED** [review]

File: src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  public FileResourceDTO downloadExportFile(Long organizationId,
+    Long exportFileId, UserInfo user,
+    String accessToken) throws FileNotFoundException {
+    userAuthorizationService.checkUserAuthorization(organizationId, user,
+      accessToken);
+
+    ExportFile exportFile = exportFileService.getExportFile(exportFileId, accessToken);
+
+    if (exportFile == null) {
+      throw new FileNotFoundException("Export file with id %s was not found".formatted(exportFileId));
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/service/export/ExportFileFacadeServiceImpl.java]: nope! create a new custom class if it doesn't exists, then add it's translation into the ControllerAdvice class and add the test on the related Test class

**Review by antonioT90 — CHANGES_REQUESTED** [review]
Please, fix also Sonar warnings

File: src/main/java/it/gov/pagopa/pu/fileshare/exception/custom/FlowFileNotFoundException.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package it.gov.pagopa.pu.fileshare.exception.custom;
+
+public class FlowFileNotFoundException extends RuntimeException {
```
Comment (antonioT90) [inline src/main/java/it/gov/pagopa/pu/fileshare/exception/custom/FlowFileNotFoundException.java]: what's a FlowFile? :D 
```suggestion
public class FileNotFoundException extends RuntimeException {
```

Comment (antonioT90) [reply src/main/java/it/gov/pagopa/pu/fileshare/exception/custom/FlowFileNotFoundException.java]: ps: can you add it's throws also in ingestionFlowFile donwload?

**Review by antonioT90 — CHANGES_REQUESTED** [review]

**Review by antonioT90 — APPROVED** [review]
